### DISCUSSION TITLE
ORM Benchmarks

### DISCUSSION THREAD

**Discussion by mikependon** [TwentyFourMinutes/Venflow#15] [Show and tell]
Hey mate, nice ORM and congrats to this.

In relation to the benchmark, RepoDB requires an initial materialization (or startup) as the first compilation is hitting some degree of performance-degradation, so it is impacting the average a lot. We also place that on our [README](https://github.com/mikependon/RepoDB#important-note).

![image](https://user-images.githubusercontent.com/591015/132311448-f79635da-b7fe-4c1f-9c09-f72f84fabf91.png)

On this link: https://github.com/TwentyFourMinutes/Venflow/blob/54d81e688cb51f454668b8af1b8f50122a80eae3/src/Venflow/Venflow.Benchmarks/Benchmarks/QueryBenchmarks/QueryBatchAsyncBenchmark.cs#L90

The first execution is very slow and the 2nd execution would be fast. RepoDB has a special mechanism to cache everything based on the projected columns, therefore I expect it will never be slower than Dapper starting from the 2nd execution (since they are not doing that null-checks). On top of that, we have added more compilations on the passed-parameter which other is not doing, and that has brought RepoDB to lead the benchmar [bencher](https://github.com/FransBouma/RawDataAccessBencher/blob/master/Results/20201112_net5_ef5.txt) of Frans Bouma.

Lastly, RepoDB supports the raw SQL execution like Dapper, so if you care about showing the performance and efficiency, you can utilize that on this materialization.

Looking forward to the further development of this ORM. 🚀

**Comment by TwentyFourMinutes** [comment]
Thanks a lot for the feedback!

About the benchmarking itself, BenchmarkDotNet itself runs each benchmark multiple times before starting to count (warmup). However to ensure that _all_ ORM's get to build their cache independently of what BenchmarkDotNet is doing. See this: https://github.com/TwentyFourMinutes/Venflow/blob/54d81e688cb51f454668b8af1b8f50122a80eae3/src/Venflow/Venflow.Benchmarks/Benchmarks/QueryBenchmarks/QueryBatchAsyncBenchmark.cs#L47

Nonetheless I'd would be more than happy to change it to raw SQL for RepoDb as well, as I am assuming best case scenarios for all ORM's. So if you could help me on that, it would be great!

On another note, I am not quite sure what RawDataAccessBencher is actually measuring, as I never looked into it. However, it is quite interesting to see such a huge difference between the RepoDb results and the ones from Dapper, especially for SingleQueries. Any ideas?

**Reply by mikependon** [reply]
Dapper is known to be fast as well in [single](https://github.com/FransBouma/RawDataAccessBencher/blob/716e774dd5664ebad11a01cf45c498a74e134326/Results/20201112_net5_ef5.txt#L106) query, but is quite slow in [set](https://github.com/FransBouma/RawDataAccessBencher/blob/716e774dd5664ebad11a01cf45c498a74e134326/Results/20201112_net5_ef5.txt#L60) query. Some logics we implemented during the AOT compilation (to both RepoDB and LLBLGen) is absent in Dapper, that's the reason why it will never be faster.

The reason why RepoDB is ahead of the bencher is because of the additional caching layer we implemented on the anonymous object parameters.

**Reply by TwentyFourMinutes** [reply]
That's quite interesting!
